generators/typescript/codegen/src/ast/core/Writer.ts

```diff
@@ -1,7 +1,157 @@
+
+type ModuleName = string;
+type Alias = string;
+
+export declare namespace Writer {
+    interface Args {
+        /* The name of the module that is being written to */
+        moduleName: string;
+    }
+}
 
 export class Writer extends AbstractWriter {
+    protected imports: Record<ModuleName, Reference[]> = {};
+    protected defaultImports: Record<ModuleName, Reference> = {};
+    protected starImportAliases: Record<ModuleName, Alias> = {};
+    protected starImportAliasesInverse: Record<Alias, ModuleName> = {};
+
+    /**
+     * Adds the given import under its module name.
+     */
+    public addImport(reference: Reference): void {
+        if (reference.importFrom != null) {
+            switch (reference.importFrom.type) {
+                case "default": {
+                    this.validateDefault(reference);
+                    break;
+                }
+                case "named": {
+                    this.validateNamed(reference);
+                    break;
+                }
+                case "star": {
+                    this.validateStar(reference);
+                    break;
+                }
+            }
+            const moduleImports = (this.imports[reference.importFrom.moduleName] ??= []);
+            const names = moduleImports.map((import_) => import_.name);
+            if (!names.includes(reference.name)) {
+                moduleImports.push(reference);
+            }
+        }
+    }
+
+    private validateDefault(reference: Reference): void {
+        if (reference.importFrom?.type !== "default") {
+            return;
+        }
+        const moduleDefault = (this.defaultImports[reference.importFrom.moduleName] ??= reference);
+        if (moduleDefault.name !== reference.name) {
+            throw new Error(
+                `Cannot have multiple default imports for module ${reference.importFrom.moduleName}: ` +
+                    `got ${reference.name} but already had ${moduleDefault.name}`
+            );
+        }
+    }
+
+    private validateNamed(reference: Reference): void {
+        if (reference.importFrom?.type !== "named") {
+            return;
+        }
+
+        const existing = this.imports[reference.importFrom.moduleName] ?? [];
+        const existingStar = existing.filter((e) => e.importFrom?.type === "star");
+        if (existingStar.length > 0) {
+            throw new Error(
+                `Cannot add named import ${reference.name} because non-named` +
+                    ` imports ${existingStar.map((e) => e.name)} already exist`
+            );
+        }
+
+        const duplicates = [];
+        for (const references of Object.values(this.imports)) {
+            for (const ref of references) {
+                if (ref.importFrom?.type === "named" && ref.name === reference.name) {
+                    duplicates.push(ref);
+                }
+            }
+        }
+        // TODO: Need to be able to resolve conflicts instead of throwing
```

Comment:
Discussed offline - 

We'll need to build this so that it's resilient to import conflicts. For example, consider the following:

```typescript
export { File } from "./acme/foo/File";
export { File } from "./acme/bar/File";
```

We don't want to throw in these cases - the generator should handle the alias and assign it as needed.

```typescript
export { File } from "./acme/foo/File";
export { File as File_ } from "./acme/bar/File";
```